sys/hashes/sha1.c

```diff
@@ -0,0 +1,164 @@
+        if (i >= 16) {
+            t = s->buffer[(i + 13) & 15] ^ s->buffer[(i + 8) & 15] ^
+                s->buffer[(i + 2) & 15] ^ s->buffer[i & 15];
+            s->buffer[i & 15] = sha1_rol32(t, 1);
+        }
+        if (i < 20) {
+            t = (d ^ (b & (c ^ d))) + SHA1_K0;
+        }
+        else if (i < 40) {
+            t = (b ^ c ^ d) + SHA1_K20;
+        }
+        else if (i < 60) {
+            t = ((b & c) | (d & (b | c))) + SHA1_K40;
+        }
+        else {
+            t = (b ^ c ^ d) + SHA1_K60;
+        }
+        t += sha1_rol32(a, 5) + e + s->buffer[i & 15];
+        e = d;
+        d = c;
+        c = sha1_rol32(b, 30);
+        b = a;
+        a = t;
+    }
+    s->state[0] += a;
+    s->state[1] += b;
+    s->state[2] += c;
+    s->state[3] += d;
+    s->state[4] += e;
+}
+
+void sha1_add_uncounted(sha1_context *s, uint8_t data)
+{
+    uint8_t *const b = (uint8_t *) s->buffer;
+
+#ifdef __BIG_ENDIAN__
+    b[s->buffer_offset] = data;
+#else
+    b[s->buffer_offset ^ 3] = data;
+#endif
+    s->buffer_offset++;
+    if (s->buffer_offset == SHA1_BLOCK_LENGTH) {
+        sha1_hash_block(s);
+        s->buffer_offset = 0;
+    }
+}
+
+void sha1_update(sha1_context *s, const unsigned char *data, size_t len)
+{
+    while (len--) {
+        ++s->byte_count;
+        sha1_add_uncounted(s, *data++);
+    }
+}
+
+void sha1_pad(sha1_context *s)
+{
+    /* Implement SHA-1 padding (fips180-2 §5.1.1) */
+
+    /* Pad with 0x80 followed by 0x00 until the end of the block */
+    sha1_add_uncounted(s, 0x80);
+    while (s->buffer_offset != 56) {
+        sha1_add_uncounted(s, 0x00);
+    }
+
+    /* Append length in the last 8 bytes */
+    sha1_add_uncounted(s, 0);                   /* We're only using 32 bit lengths */
+    sha1_add_uncounted(s, 0);                   /* But SHA-1 supports 64 bit lengths */
+    sha1_add_uncounted(s, 0);                   /* So zero pad the top bits */
+    sha1_add_uncounted(s, s->byte_count >> 29); /* Shifting to multiply by 8 */
+    sha1_add_uncounted(s, s->byte_count >> 21); /* as SHA-1 supports bitstreams as well as */
+    sha1_add_uncounted(s, s->byte_count >> 13); /* byte. */
+    sha1_add_uncounted(s, s->byte_count >> 5);
+    sha1_add_uncounted(s, s->byte_count << 3);
+}
+
+uint8_t *sha1_final(sha1_context *s)
+{
+    /* Pad to complete the last block */
+    sha1_pad(s);
+
+    /* Swap byte order back */
+    for (int i = 0; i < 5; i++) {
+        s->state[i] =
+            (((s->state[i]) << 24) & 0xff000000)
+            | (((s->state[i]) << 8) & 0x00ff0000)
+            | (((s->state[i]) >> 8) & 0x0000ff00)
+            | (((s->state[i]) >> 24) & 0x000000ff);
+    }
+
+    /* Return pointer to hash (20 characters) */
+    return (uint8_t *) s->state;
+}
+
+void sha1(uint8_t *dst, const uint8_t *src, size_t len)
+{
+    sha1_context ctx;
+    sha1_init(&ctx);
+    sha1_update(&ctx, (unsigned char*) src, len);
+    memcpy(dst, sha1_final(&ctx), SHA1_DIGEST_LENGTH);
```

Comment:
This works since the digest is pushed on the stack, but I'm a bit nervous looking at this